### ISSUE TITLE
Optimized and Unoptimized are not Networks Equivalent

### ISSUE THREAD

**Issue by lucasBertola** [cazala/synaptic#78]
Hi,
Create an Optimized or an Unoptimized network are not Equivalent.

For exemple, if you run the test "_Optimized and Unoptimized Networks Equivalency_" with  var iterations = 10000; we will can see that it's absolutely not Equivalent after long time process .
It's a serious problem.
I tried to solve but not success yet.
Because I do not understand enough the optimize system.
MayBe someone change Unoptimized system without change Optimized system?

Thx

**Comment by lucasBertola** [comment]
If it can help, It's not equivalent only when there are Forgetgates in network.

**Comment by cazala** [comment]
Hi @lucasBertola, I've been playing with it and you are right, outputs between optimized and unoptimized networks aren't equal, I noticed that they start to differ only when calling .propagate(), but activate() alone returns always the same output no matter how many times it's called. So there's a difference between the optimzed propagate and the unoptimized one. Also I noticed it starts to happen after the second call to propagate(). This makes more sense after what you said that only happens when there are gated connections, because it probably has to do with the computation of the extended traces, that's why it only happens after the second time and not the first time propagate() is called (the first time there are no extended traces). I'll have to debug step by step the propagation of both networks until finding where's the difference.

**Comment by cazala** [comment]
btw, is it ONLY Forget gates, or any kind of gates?

**Comment by lucasBertola** [comment]
Yes, I think it's only with the forget gates, but I am not 100% sure.
When I try to manipulate different Gates in architec.js; it's only with Forget gates that I obtain some errors of equivalence.

**Comment by cazala** [comment]
Nice, that narrows the surface of code to debug a lot, since forget gates only gate self-connections, there a tiny slice of the code that takes care of that special case. I'll give it a look.

**Comment by lucasBertola** [comment]
I tested like you to remove the propagate function. And I have still errors. ( I think you should use "`output1- output2 != 0`" to compare , and not "`compare()`".

So I think the probleme come from activate.

**Comment by cazala** [comment]
The thing is, since the optimized function uses a Float64Array, instead of Number objects, there might be tiny differences in the outputs due to truncation. You can check this by printing the substraction of the outputs, like you said. You will notice that most of the times they are 0, but sometimes they are very little like 1e-15. That amount of error is acceptable, but it will break your assertions if you make them like "a - b != 0"

**Comment by lucasBertola** [comment]
I understand. But, how is it possible that I never had any error with Perceptron ( `var optimized = new Perceptron(2,1,1) `) with `output1- output2 != 0` ?

**Comment by lucasBertola** [comment]
And the truncation should be same for all equivalent network, no? So I really think that `output1- output2 != 0` should work

**Comment by lucasBertola** [comment]
Hi,
I think I found... I'll keep looking. I'll keep you informed

**Comment by cazala** [comment]
awesome! I still haven't found time to sit and debug this, but keep me posted if you find anything :)

**Comment by lucasBertola** [comment]
I sent a pull Request for repair `Activate `Method.

But `propagate` method still doesn't work.

**Comment by cazala** [comment]
I wrote comments on the PR
